File: sdk-ts/examples/quickstart.ts
/**
 * End-to-end TypeScript example against a `veilux serve` dev node.
 *
 *   veilux serve --addr 127.0.0.1:8645
 *   node --loader ts-node/esm examples/quickstart.ts
 *
 * Or after `npm run build`, compile this file and run with node.
 */
import { Client, PartyIdentity, builders, hashCommit, toHex } from "../dist/index.js";

const endpoint = process.env.VEILUX_RPC ?? "http://127.0.0.1:8645";

async function main() {
  const client = new Client(endpoint);

  // Deterministic identity: 32-byte seed of all 1s (matches Rust [1u8;32]).
  const seed = new Uint8Array(32).fill(1);
  const alice = PartyIdentity.fromSeed("alice", seed);

  const info = await client.nodeInfo();
  console.log(`node: network=${info.network} height=${info.height} prisms=${info.prisms.join(",")}`);

  // 1) Create a token.
  const create = builders.tokenCreate("alice", "Public", 0, "Gold Coin", "GLD", 18, 1_000_000n, true);
  const est = await client.estimate(alice.sign(create));
  console.log(`estimated cost: ${est.cost} LUX`);

  const res = await client.submit(alice.sign(create));
  console.log(`token create accepted=${res.accepted} id=${res.command_id}`);

  // 2) Compute the token id the same way the node does, then transfer.
  const te = new TextEncoder();
  const tokenId = toHex(
    hashCommit("token/id", [te.encode("alice"), te.encode("GLD"), te.encode("Gold Coin")]),
  );
  const transfer = builders.tokenTransfer("alice", "Public", 1, tokenId, "bob", 250_000n);
  const res2 = await client.submit(alice.sign(transfer));
  console.log(`transfer accepted=${res2.accepted}`);

  // 3) Read state back.
  const height = await client.blockNumber();
  const block = await client.blockByNumber(height);
  console.log(`head block #${block.height} hash=${block.hash} commands=${block.command_count}`);

  const balKey = `token/bal/${tokenId}/bob`;
  const bal = await client.getState(balKey);
  if (bal.found) {
    const decimal = Buffer.from(bal.value_hex, "hex").toString("utf8");
    console.log(`bob's GLD balance: ${decimal}`);
  } else {
    console.log(`bob's GLD balance: 0 (not found)`);
  }
}

main().catch((e) => {
  console.error("error:", e);
  process.exit(1);
});

File: sdk/examples/quickstart.rs
//! End-to-end SDK example: connect to a `veilux serve` dev node, create a
//! token, transfer it, and read the chain back.
//!
//! Run a node first:  veilux serve --addr 127.0.0.1:8645
//! Then:              cargo run -p veilux-sdk --example quickstart

use veilux_sdk::{builders, Client, PartyIdentity, Visibility};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let endpoint =
        std::env::var("VEILUX_RPC").unwrap_or_else(|_| "http://127.0.0.1:8645".to_string());
    let client = Client::new(&endpoint);

    let alice = PartyIdentity::from_seed("alice", &[1u8; 32]);
    println!("connected to {endpoint}");

    let info = client.node_info()?;
    println!(
        "node: network={} height={} prisms={:?}",
        info.network, info.height, info.prisms
    );

    // 1) Create a fungible token.
    let create = builders::token_create(
        alice.party().clone(),
        Visibility::Public,
        0,
        "Gold Coin",
        "GLD",
        18,
        1_000_000,
        true,
    );
    let est = client.estimate(&alice.sign(create.clone()))?;
    println!("estimated cost: {} LUX", est.cost);

    let res = client.submit(&alice.sign(create))?;
    println!(
        "token create accepted={} id={}",
        res.accepted, res.command_id
    );

    // 2) Transfer some to bob (token id is derived deterministically).
    let token_id = veilux_sdk::Hash::commit("token/id", &[b"alice", b"GLD", b"Gold Coin"]);
    let transfer = builders::token_transfer(
        alice.party().clone(),
        Visibility::Public,
        1,
        token_id,
        veilux_sdk::PartyId::new("bob"),
        250_000,
    );
    let res = client.submit(&alice.sign(transfer))?;
    println!("transfer accepted={}", res.accepted);

    // 3) Read the chain back.
    let height = client.block_number()?;
    println!("chain height now: {height}");
    let block = client.block_by_number(height)?;
    println!(
        "head block #{} hash={} commands={}",
        block.height, block.hash, block.command_count
    );

    // 4) Read bob's token balance from state (stored as a decimal string).
    let bal_key = format!("token/bal/{}/bob", token_id.to_hex());
    let bal = client.get_state(&bal_key)?;
    if bal.found {
        let decimal =
            String::from_utf8(hex::decode(&bal.value_hex).unwrap_or_default()).unwrap_or_default();
        println!("bob's GLD balance: {decimal}");
    } else {
        println!("bob's GLD balance: 0 (key {bal_key} not found)");
    }

    Ok(())
}
